File: lib/user/validate_point_config.dart
/// A-7: ポイント関連 config 整合性 validation（Flutter）
///
/// 必須設定の欠損を default で補完しない。
/// Functions の validatePointConfig と同等の判定を返す。
library;

import 'point_ids.dart';
import 'point_conversion.dart';

const String kConfigPointInvalid = 'CONFIG_POINT_INVALID';
const int kDisplayNameMaxLength = 40;

/// bill 会計で使う既知カテゴリ（Functions の KNOWN_BILL_CATEGORIES と同一）
const List<String> kKnownBillCategories = [
  'extraCost',
  'sideGameChip',
  'tournaments',
  'items',
];

class PointSlotSetting {
  final bool enabled;
  final String displayName;

  const PointSlotSetting({
    required this.enabled,
    required this.displayName,
  });
}

class SideGameChipSettings {
  final bool enabled;
  final String displayName;

  const SideGameChipSettings({
    required this.enabled,
    required this.displayName,
  });
}

class BalancePaymentSetting {
  final BalanceConversion conversion;
  final int usageUnit;

  const BalancePaymentSetting({
    required this.conversion,
    required this.usageUnit,
  });
}

class ValidatedPointConfig {
  final Map<String, PointSlotSetting> pointSettings;
  final SideGameChipSettings sideGameChipSettings;
  final List<String> rankingRewardPointTypes;
  final Map<String, List<String>> categoryPaymentMethods;
  final List<String> pointPriority;
  final Map<String, BalancePaymentSetting> balancePaymentSettings;
  final List<String> categoryOrder;

  const ValidatedPointConfig({
    required this.pointSettings,
    required this.sideGameChipSettings,
    required this.rankingRewardPointTypes,
    required this.categoryPaymentMethods,
    required this.pointPriority,
    required this.balancePaymentSettings,
    required this.categoryOrder,
  });
}

class PointConfigValidationResult {
  final bool ok;
  final ValidatedPointConfig? value;
  final String? errorKey;
  final String? message;

  const PointConfigValidationResult._({
    required this.ok,
    this.value,
    this.errorKey,
    this.message,
  });

  factory PointConfigValidationResult.success(ValidatedPointConfig value) =>
      PointConfigValidationResult._(ok: true, value: value);

  factory PointConfigValidationResult.failure(String message) =>
      PointConfigValidationResult._(
        ok: false,
        errorKey: kConfigPointInvalid,
        message: message,
      );
}

PointConfigValidationResult tryValidatePointConfig({
  required Object? pointSettings,
  required Object? sideGameChipSettings,
  required Object? rankingRewardPointTypes,
  required Object? categoryPaymentMethods,
  required Object? pointPriority,
  required Object? balancePaymentSettings,
  required Object? categoryOrder,
}) {
  try {
    final ps = _validatePointSettings(pointSettings);
    final chip = _validateSideGameChipSettings(sideGameChipSettings);
    final rewards = _validateRankingRewardPointTypes(
      rankingRewardPointTypes,
      ps,
    );
    final categories = _validateCategoryPaymentMethods(categoryPaymentMethods);
    _validateEnabledVsAllowlists(ps, chip, categories);
    final payable = _collectPayableBalanceIds(categories);
    final balanceSettings = _validateBalancePaymentSettings(
      balancePaymentSettings,
      payable,
    );
    final priority = _validatePointPriority(pointPriority, ps, chip, payable);
    final order = _validateCategoryOrder(categoryOrder);
    return PointConfigValidationResult.success(
      ValidatedPointConfig(
        pointSettings: ps,
        sideGameChipSettings: chip,
        rankingRewardPointTypes: rewards,
        categoryPaymentMethods: categories,
        pointPriority: priority,
        balancePaymentSettings: balanceSettings,
        categoryOrder: order,
      ),
    );
  } on _PointConfigException catch (e) {
    return PointConfigValidationResult.failure(e.message);
  }
}

class _PointConfigException implements Exception {
  final String message;
  _PointConfigException(this.message);
}

Never _reject(String message) => throw _PointConfigException(message);

Map<String, PointSlotSetting> _validatePointSettings(Object? raw) {
  if (raw is! Map) {
    _reject('pointSettings が存在しないか不正です');
  }
  final out = <String, PointSlotSetting>{};
  for (final id in kCurrencyPointIds) {
    if (!raw.containsKey(id)) {
      _reject('pointSettings.$id が欠落しています');
    }
    out[id] = _validateSlotSetting(raw[id], 'pointSettings.$id');
  }
  for (final key in raw.keys) {
    if (!isCurrencyPointId(key.toString())) {
      _reject('pointSettings に未知のキーがあります: $key');
    }
  }
  return out;
}

SideGameChipSettings _validateSideGameChipSettings(Object? raw) {
  final slot = _validateSlotSetting(raw, 'sideGameChipSettings');
  return SideGameChipSettings(
    enabled: slot.enabled,
    displayName: slot.displayName,
  );
}

PointSlotSetting _validateSlotSetting(Object? raw, String path) {
  if (raw is! Map) {
    _reject('$path は object である必要があります');
  }
  final enabled = raw['enabled'];
  if (enabled is! bool) {
    _reject('$path.enabled は boolean である必要があります');
  }
  final displayNameRaw = raw['displayName'];
  if (displayNameRaw is! String) {
    _reject('$path.displayName は string である必要があります');
  }
  final trimmed = displayNameRaw.trim();
  if (trimmed.isEmpty) {
    _reject('$path.displayName は trim 後 1 文字以上である必要があります');
  }
  if (trimmed.length > kDisplayNameMaxLength) {
    _reject('$path.displayName は最大 $kDisplayNameMaxLength 文字です');
  }
  return PointSlotSetting(enabled: enabled, displayName: trimmed);
}

List<String> _validateRankingRewardPointTypes(
  Object? raw,
  Map<String, PointSlotSetting> pointSettings,
) {
  if (raw == null) {
    _reject('tournament.rankingRewardPointTypes が欠落しています');
  }
  if (raw is! List) {
    _reject('rankingRewardPointTypes は配列である必要があります');
  }
  final seen = <String>{};
  final out = <String>[];
  for (final item in raw) {
    final id = item?.toString();
    if (id == kSideGameChipId) {
      _reject('rankingRewardPointTypes に sideGameChip は含められません');
    }
    if (!isCurrencyPointId(id)) {
      _reject('rankingRewardPointTypes に未知または不正な ID: $item');
    }
    if (seen.contains(id)) {
      _reject('rankingRewardPointTypes に重複があります: $id');
    }
    if (pointSettings[id]!.enabled != true) {
      _reject('rankingRewardPointTypes の $id は enabled:false です');
    }
    seen.add(id!);
    out.add(id);
  }
  return out;
}

Map<String, List<String>> _validateCategoryPaymentMethods(Object? raw) {
  if (raw is! Map) {
    _reject('categoryPaymentMethods が存在しないか不正です');
  }
  final out = <String, List<String>>{};
  for (final entry in raw.entries) {
    final category = entry.key.toString();
    final methodsRaw = entry.value;
    if (methodsRaw is! List) {
      _reject('categoryPaymentMethods.$category は配列である必要があります');
    }
    final methods = <String>[];
    for (final method in methodsRaw) {
      final m = method?.toString();
      if (m == null || (!isCashLikeMethod(m) && !isBalanceId(m))) {
        _reject('categoryPaymentMethods.$category に未知の method: $method');
      }
      methods.add(m);
    }
    out[category] = methods;
  }
  return out;
}

Set<String> _collectPayableBalanceIds(
  Map<String, List<String>> categoryPaymentMethods,
) {
  final set = <String>{};
  for (final methods in categoryPaymentMethods.values) {
    for (final method in methods) {
      if (isBalanceId(method)) set.add(method);
    }
  }
  return set;
}

bool _isBalanceEnabled(
  String id,
  Map<String, PointSlotSetting> pointSettings,
  SideGameChipSettings chip,
) {
  if (id == kSideGameChipId) return chip.enabled;
  return pointSettings[id]?.enabled == true;
}

void _validateEnabledVsAllowlists(
  Map<String, PointSlotSetting> pointSettings,
  SideGameChipSettings chip,
  Map<String, List<String>> categoryPaymentMethods,
) {
  for (final entry in categoryPaymentMethods.entries) {
    for (final method in entry.value) {
      if (!isBalanceId(method)) continue;
      if (!_isBalanceEnabled(method, pointSettings, chip)) {
        _reject(
          'categoryPaymentMethods.${entry.key} の $method は enabled:false です',
        );
      }
    }
  }
}

Map<String, BalancePaymentSetting> _validateBalancePaymentSettings(
  Object? raw,
  Set<String> payableBalanceIds,
) {
  if (raw is! Map) {
    _reject('balancePaymentSettings が存在しないか不正です');
  }
  final out = <String, BalancePaymentSetting>{};
  for (final entry in raw.entries) {
    final key = entry.key.toString();
    if (!isBalanceId(key)) {
      _reject('balancePaymentSettings に未知の ID: $key');
    }
    out[key] = _validateBalancePaymentSetting(
      entry.value,
      'balancePaymentSettings.$key',
    );
  }
  for (final id in payableBalanceIds) {
    if (!out.containsKey(id)) {
      _reject(
        'categoryPaymentMethods に含まれる $id の balancePaymentSettings がありません',
      );
    }
  }
  return out;
}

BalancePaymentSetting _validateBalancePaymentSetting(
  Object? raw,
  String path,
) {
  if (raw is! Map) {
    _reject('$path は object である必要があります');
  }
  final conversionRaw = raw['conversion'];
  if (conversionRaw is! Map) {
    _reject('$path.conversion は object である必要があります');
  }
  final referenceUnits = conversionRaw['referenceUnits'];
  final balanceUnits = conversionRaw['balanceUnits'];
  final usageUnit = raw['usageUnit'];
  if (!_isPositiveSafeInt(referenceUnits)) {
    _reject('$path.conversion.referenceUnits は正の安全整数である必要があります');
  }
  if (!_isPositiveSafeInt(balanceUnits)) {
    _reject('$path.conversion.balanceUnits は正の安全整数である必要があります');
  }
  if (!_isPositiveSafeInt(usageUnit)) {
    _reject('$path.usageUnit は正の安全整数である必要があります');
  }
  return BalancePaymentSetting(
    conversion: BalanceConversion(
      referenceUnits: referenceUnits as int,
      balanceUnits: balanceUnits as int,
    ),
    usageUnit: usageUnit as int,
  );
}

/// categoryOrder は config 正本（ハードコード順 fallback 禁止）。
List<String> _validateCategoryOrder(Object? raw) {
  if (raw == null) {
    _reject('categoryOrder が欠落しています');
  }
  if (raw is! List || raw.isEmpty) {
    _reject('categoryOrder は非空の配列である必要があります');
  }
  final known = kKnownBillCategories.toSet();
  final seen = <String>{};
  final out = <String>[];
  for (final item in raw) {
    final id = item?.toString();
    if (id == null || !known.contains(id)) {
      _reject('categoryOrder に未知のカテゴリがあります: $item');
    }
    if (seen.contains(id)) {
      _reject('categoryOrder に重複があります: $id');
    }
    seen.add(id);
    out.add(id);
  }
  for (final required in kKnownBillCategories) {
    if (!seen.contains(required)) {
      _reject('categoryOrder に必須カテゴリ $required がありません');
    }
  }
  return out;
}

bool _isPositiveSafeInt(Object? n) {
  if (n is! int) return false;
  return n > 0 && n <= kMaxSafeInteger;
}

List<String> _validatePointPriority(
  Object? raw,
  Map<String, PointSlotSetting> pointSettings,
  SideGameChipSettings chip,
  Set<String> payableBalanceIds,
) {
  if (raw == null) {
    _reject('pointPriority が欠落しています');
  }
  if (raw is! List) {
    _reject('pointPriority は配列である必要があります');
  }
  final seen = <String>{};
  final out = <String>[];
  for (final item in raw) {
    final id = item?.toString();
    if (isCashLikeMethod(id)) {
      _reject('pointPriority に現金系 method は含められません: $id');
    }
    if (!isBalanceId(id)) {
      _reject('pointPriority に未知の ID: $item');
    }
    if (seen.contains(id)) {
      _reject('pointPriority に重複があります: $id');
    }
    if (!_isBalanceEnabled(id!, pointSettings, chip)) {
      _reject('pointPriority の $id は enabled:false です');
    }
    if (!payableBalanceIds.contains(id)) {
      _reject(
        'pointPriority の $id は categoryPaymentMethods 上の支払可能残高ではありません',
      );
    }
    seen.add(id);
    out.add(id);
  }
  return out;
}

File: functions/src/shared/config/validatePointConfig.ts
/**
 * A-7: ポイント関連 config の整合性 validation
 *
 * 必須設定の欠損を旧値 / default で補完しない。
 * 業務不備は FunctionCustomError(CONFIG_POINT_INVALID)。
 */

import { FunctionCustomError } from '../logging/functionCustomError';
import {
  BalanceId,
  CURRENCY_POINT_IDS,
  CurrencyPointId,
  isBalanceId,
  isCashLikeMethod,
  isCurrencyPointId,
  SIDE_GAME_CHIP_ID,
} from '../../domains/user/types/pointIds';
import type {
  BalancePaymentSetting,
  BalancePaymentSettings,
  PointSettings,
  PointSlotSetting,
  SideGameChipSettings,
  StoreConfig,
} from './types';

export const CONFIG_POINT_INVALID = 'CONFIG_POINT_INVALID' as const;

export const DISPLAY_NAME_MAX_LENGTH = 40;

/** validation 入力（StoreConfig 断片または raw） */
export type PointConfigValidationInput = {
  pointSettings?: unknown;
  sideGameChipSettings?: unknown;
  rankingRewardPointTypes?: unknown;
  categoryPaymentMethods?: unknown;
  pointPriority?: unknown;
  balancePaymentSettings?: unknown;
  categoryOrder?: unknown;
};

export type ValidatedPointConfig = {
  pointSettings: PointSettings;
  sideGameChipSettings: SideGameChipSettings;
  rankingRewardPointTypes: CurrencyPointId[];
  categoryPaymentMethods: Record<string, string[]>;
  pointPriority: BalanceId[];
  balancePaymentSettings: BalancePaymentSettings;
  categoryOrder: string[];
};

/** bill 会計で使う既知カテゴリ */
export const KNOWN_BILL_CATEGORIES = [
  'extraCost',
  'sideGameChip',
  'tournaments',
  'items',
] as const;

function reject(message: string, context?: Record<string, unknown>): never {
  throw new FunctionCustomError({
    errorKey: CONFIG_POINT_INVALID,
    message,
    context,
  });
}

function isPositiveSafeInteger(n: unknown): n is number {
  return (
    typeof n === 'number' &&
    Number.isInteger(n) &&
    n > 0 &&
    n <= Number.MAX_SAFE_INTEGER
  );
}

function validateDisplayName(raw: unknown, path: string): string {
  if (typeof raw !== 'string') {
    reject(`${path}.displayName は string である必要があります`, { path });
  }
  const trimmed = raw.trim();
  if (trimmed.length < 1) {
    reject(`${path}.displayName は trim 後 1 文字以上である必要があります`, {
      path,
    });
  }
  if (trimmed.length > DISPLAY_NAME_MAX_LENGTH) {
    reject(
      `${path}.displayName は最大 ${DISPLAY_NAME_MAX_LENGTH} 文字です`,
      { path, length: trimmed.length },
    );
  }
  return trimmed;
}

function validateSlotSetting(raw: unknown, path: string): PointSlotSetting {
  if (raw == null || typeof raw !== 'object' || Array.isArray(raw)) {
    reject(`${path} は object である必要があります`, { path });
  }
  const obj = raw as Record<string, unknown>;
  if (typeof obj.enabled !== 'boolean') {
    reject(`${path}.enabled は boolean である必要があります`, { path });
  }
  return {
    enabled: obj.enabled,
    displayName: validateDisplayName(obj.displayName, path),
  };
}

function validatePointSettings(raw: unknown): PointSettings {
  if (raw == null || typeof raw !== 'object' || Array.isArray(raw)) {
    reject('pointSettings が存在しないか不正です');
  }
  const obj = raw as Record<string, unknown>;
  const out = {} as PointSettings;
  for (const id of CURRENCY_POINT_IDS) {
    if (!Object.prototype.hasOwnProperty.call(obj, id)) {
      reject(`pointSettings.${id} が欠落しています`, { id });
    }
    out[id] = validateSlotSetting(obj[id], `pointSettings.${id}`);
  }
  for (const key of Object.keys(obj)) {
    if (!isCurrencyPointId(key)) {
      reject(`pointSettings に未知のキーがあります: ${key}`, { key });
    }
  }
  return out;
}

function validateSideGameChipSettings(raw: unknown): SideGameChipSettings {
  if (raw == null || typeof raw !== 'object' || Array.isArray(raw)) {
    reject('sideGameChipSettings が存在しないか不正です');
  }
  return validateSlotSetting(raw, 'sideGameChipSettings');
}

function isBalanceEnabled(
  id: BalanceId,
  pointSettings: PointSettings,
  sideGameChipSettings: SideGameChipSettings,
): boolean {
  if (id === SIDE_GAME_CHIP_ID) {
    return sideGameChipSettings.enabled;
  }
  return pointSettings[id].enabled;
}

function validateRankingRewardPointTypes(
  raw: unknown,
  pointSettings: PointSettings,
): CurrencyPointId[] {
  if (raw === undefined) {
    reject('tournament.rankingRewardPointTypes が欠落しています');
  }
  if (!Array.isArray(raw)) {
    reject('rankingRewardPointTypes は配列である必要があります');
  }
  const seen = new Set<string>();
  const out: CurrencyPointId[] = [];
  for (const item of raw) {
    if (item === SIDE_GAME_CHIP_ID) {
      reject('rankingRewardPointTypes に sideGameChip は含められません');
    }
    if (!isCurrencyPointId(item)) {
      reject(`rankingRewardPointTypes に未知または不正な ID: ${String(item)}`, {
        item,
      });
    }
    if (seen.has(item)) {
      reject(`rankingRewardPointTypes に重複があります: ${item}`, { item });
    }
    if (!pointSettings[item].enabled) {
      reject(
        `rankingRewardPointTypes の ${item} は enabled:false です`,
        { item },
      );
    }
    seen.add(item);
    out.push(item);
  }
  return out;
}

function validateCategoryPaymentMethods(
  raw: unknown,
): Record<string, string[]> {
  if (raw == null || typeof raw !== 'object' || Array.isArray(raw)) {
    reject('categoryPaymentMethods が存在しないか不正です');
  }
  const obj = raw as Record<string, unknown>;
  const out: Record<string, string[]> = {};
  for (const [category, methodsRaw] of Object.entries(obj)) {
    if (!Array.isArray(methodsRaw)) {
      reject(`categoryPaymentMethods.${category} は配列である必要があります`, {
        category,
      });
    }
    const methods: string[] = [];
    for (const method of methodsRaw) {
      if (typeof method !== 'string') {
        reject(`categoryPaymentMethods.${category} の要素が string ではありません`, {
          category,
          method,
        });
      }
      if (!isCashLikeMethod(method) && !isBalanceId(method)) {
        reject(
          `categoryPaymentMethods.${category} に未知の method: ${method}`,
          { category, method },
        );
      }
      methods.push(method);
    }
    out[category] = methods;
  }
  return out;
}

function collectPayableBalanceIds(
  categoryPaymentMethods: Record<string, string[]>,
): Set<BalanceId> {
  const set = new Set<BalanceId>();
  for (const methods of Object.values(categoryPaymentMethods)) {
    for (const method of methods) {
      if (isBalanceId(method)) {
        set.add(method);
      }
    }
  }
  return set;
}

function validateBalancePaymentSetting(
  raw: unknown,
  path: string,
): BalancePaymentSetting {
  if (raw == null || typeof raw !== 'object' || Array.isArray(raw)) {
    reject(`${path} は object である必要があります`, { path });
  }
  const obj = raw as Record<string, unknown>;
  const conversionRaw = obj.conversion;
  if (
    conversionRaw == null ||
    typeof conversionRaw !== 'object' ||
    Array.isArray(conversionRaw)
  ) {
    reject(`${path}.conversion は object である必要があります`, { path });
  }
  const conversionObj = conversionRaw as Record<string, unknown>;
  const referenceUnits = conversionObj.referenceUnits;
  const balanceUnits = conversionObj.balanceUnits;
  const usageUnit = obj.usageUnit;

  if (!isPositiveSafeInteger(referenceUnits)) {
    reject(`${path}.conversion.referenceUnits は正の安全整数である必要があります`, {
      path,
      referenceUnits,
    });
  }
  if (!isPositiveSafeInteger(balanceUnits)) {
    reject(`${path}.conversion.balanceUnits は正の安全整数である必要があります`, {
      path,
      balanceUnits,
    });
  }
  if (!isPositiveSafeInteger(usageUnit)) {
    reject(`${path}.usageUnit は正の安全整数である必要があります`, {
      path,
      usageUnit,
    });
  }

  return {
    conversion: { referenceUnits, balanceUnits },
    usageUnit,
  };
}

function validateBalancePaymentSettings(
  raw: unknown,
  payableBalanceIds: Set<BalanceId>,
): BalancePaymentSettings {
  if (raw == null || typeof raw !== 'object' || Array.isArray(raw)) {
    reject('balancePaymentSettings が存在しないか不正です');
  }
  const obj = raw as Record<string, unknown>;
  const out: BalancePaymentSettings = {};

  for (const key of Object.keys(obj)) {
    if (!isBalanceId(key)) {
      reject(`balancePaymentSettings に未知の ID: ${key}`, { key });
    }
    out[key] = validateBalancePaymentSetting(
      obj[key],
      `balancePaymentSettings.${key}`,
    );
  }

  for (const id of payableBalanceIds) {
    if (out[id] == null) {
      reject(
        `categoryPaymentMethods に含まれる ${id} の balancePaymentSettings がありません`,
        { id },
      );
    }
  }

  return out;
}

function validatePointPriority(
  raw: unknown,
  pointSettings: PointSettings,
  sideGameChipSettings: SideGameChipSettings,
  payableBalanceIds: Set<BalanceId>,
): BalanceId[] {
  if (raw === undefined) {
    reject('pointPriority が欠落しています');
  }
  if (!Array.isArray(raw)) {
    reject('pointPriority は配列である必要があります');
  }
  const seen = new Set<string>();
  const out: BalanceId[] = [];
  for (const item of raw) {
    if (isCashLikeMethod(item)) {
      reject(`pointPriority に現金系 method は含められません: ${item}`, {
        item,
      });
    }
    if (!isBalanceId(item)) {
      reject(`pointPriority に未知の ID: ${String(item)}`, { item });
    }
    if (seen.has(item)) {
      reject(`pointPriority に重複があります: ${item}`, { item });
    }
    if (!isBalanceEnabled(item, pointSettings, sideGameChipSettings)) {
      reject(`pointPriority の ${item} は enabled:false です`, { item });
    }
    if (!payableBalanceIds.has(item)) {
      reject(
        `pointPriority の ${item} は categoryPaymentMethods 上の支払可能残高ではありません`,
        { item },
      );
    }
    seen.add(item);
    out.push(item);
  }
  return out;
}

function validateEnabledVsAllowlists(
  pointSettings: PointSettings,
  sideGameChipSettings: SideGameChipSettings,
  categoryPaymentMethods: Record<string, string[]>,
): void {
  for (const [category, methods] of Object.entries(categoryPaymentMethods)) {
    for (const method of methods) {
      if (!isBalanceId(method)) continue;
      if (!isBalanceEnabled(method, pointSettings, sideGameChipSettings)) {
        reject(
          `categoryPaymentMethods.${category} の ${method} は enabled:false です`,
          { category, method },
        );
      }
    }
  }
}

function validateCategoryOrder(raw: unknown): string[] {
  if (raw === undefined) {
    reject('categoryOrder が欠落しています');
  }
  if (!Array.isArray(raw) || raw.length === 0) {
    reject('categoryOrder は非空の配列である必要があります');
  }
  const known = new Set<string>(KNOWN_BILL_CATEGORIES);
  const seen = new Set<string>();
  const out: string[] = [];
  for (const item of raw) {
    if (typeof item !== 'string' || !known.has(item)) {
      reject(`categoryOrder に未知のカテゴリがあります: ${String(item)}`, {
        item,
      });
    }
    if (seen.has(item)) {
      reject(`categoryOrder に重複があります: ${item}`, { item });
    }
    seen.add(item);
    out.push(item);
  }
  for (const required of KNOWN_BILL_CATEGORIES) {
    if (!seen.has(required)) {
      reject(`categoryOrder に必須カテゴリ ${required} がありません`, {
        required,
      });
    }
  }
  return out;
}

/**
 * A-7 ポイント config を検証し、正規化結果を返す。
 * 失敗時は FunctionCustomError(CONFIG_POINT_INVALID)。
 */
export function validatePointConfig(
  input: PointConfigValidationInput,
): ValidatedPointConfig {
  const pointSettings = validatePointSettings(input.pointSettings);
  const sideGameChipSettings = validateSideGameChipSettings(
    input.sideGameChipSettings,
  );
  const rankingRewardPointTypes = validateRankingRewardPointTypes(
    input.rankingRewardPointTypes,
    pointSettings,
  );
  const categoryPaymentMethods = validateCategoryPaymentMethods(
    input.categoryPaymentMethods,
  );
  validateEnabledVsAllowlists(
    pointSettings,
    sideGameChipSettings,
    categoryPaymentMethods,
  );
  const payableBalanceIds = collectPayableBalanceIds(categoryPaymentMethods);
  const balancePaymentSettings = validateBalancePaymentSettings(
    input.balancePaymentSettings,
    payableBalanceIds,
  );
  const pointPriority = validatePointPriority(
    input.pointPriority,
    pointSettings,
    sideGameChipSettings,
    payableBalanceIds,
  );
  const categoryOrder = validateCategoryOrder(input.categoryOrder);

  return {
    pointSettings,
    sideGameChipSettings,
    rankingRewardPointTypes,
    categoryPaymentMethods,
    pointPriority,
    balancePaymentSettings,
    categoryOrder,
  };
}

/** StoreConfig から A-7 入力を抽出して検証（Callable 直前用） */
export function validatePointConfigFromStoreConfig(
  config: StoreConfig,
): ValidatedPointConfig {
  return validatePointConfig({
    pointSettings: config.pointSettings,
    sideGameChipSettings: config.sideGameChipSettings,
    rankingRewardPointTypes: config.tournament?.rankingRewardPointTypes,
    categoryPaymentMethods: config.billing?.paymentPolicy?.categoryPaymentMethods,
    pointPriority: config.billing?.paymentPolicy?.pointPriority,
    balancePaymentSettings:
      config.billing?.paymentPolicy?.balancePaymentSettings,
    categoryOrder: config.billing?.paymentPolicy?.categoryOrder,
  });
}

/** テスト・UI 向け: throw せず結果を返す */
export function tryValidatePointConfig(
  input: PointConfigValidationInput,
):
  | { ok: true; value: ValidatedPointConfig }
  | { ok: false; errorKey: typeof CONFIG_POINT_INVALID; message: string } {
  try {
    return { ok: true, value: validatePointConfig(input) };
  } catch (e) {
    if (e instanceof FunctionCustomError && e.errorKey === CONFIG_POINT_INVALID) {
      return { ok: false, errorKey: CONFIG_POINT_INVALID, message: e.message };
    }
    throw e;
  }
}
